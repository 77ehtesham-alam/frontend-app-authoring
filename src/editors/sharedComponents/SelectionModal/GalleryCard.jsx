import React from 'react';
import PropTypes from 'prop-types';

import {
  Button,
  Icon,
  Image,
  SelectableBox,
} from '@edx/paragon';
import { FormattedMessage, FormattedDate, FormattedTime } from '@edx/frontend-platform/i18n';
import { Link } from '@edx/paragon/icons';

import messages from './messages';

export const GalleryCard = ({
  asset,
  showId,
}) => (
  <SelectableBox className="card bg-white" key={asset.externalUrl} type="radio" value={asset.id}>
    <div className="card-div d-flex flex-row flex-nowrap">
      <Image
        style={{ width: '100px', height: '100px' }}
        src={asset.externalUrl}
      />
      <div className="img-text p-3">
        <h3>{asset.displayName}</h3>
        { showId && (
          <p>
            <Button variant="link" size="inline" onClick={() => { /* TODO */ }}>
              <Icon src={Link} /> {asset.id}
            </Button>
          </p>
        )}
        <p>
          <FormattedMessage
            {...messages.addedDate}
            values={{
              date: <FormattedDate value={asset.dateAdded} />,
              time: <FormattedTime value={asset.dateAdded} />,
            }}
          />
        </p>
      </div>
    </div>
  </SelectableBox>
);

GalleryCard.defaultProps = {
  showId: false,
};

GalleryCard.propTypes = {
  asset: PropTypes.shape({
    contentType: PropTypes.string,
    displayName: PropTypes.string,
    externalUrl: PropTypes.string,
    id: PropTypes.string,
    dateAdded: PropTypes.number,
    locked: PropTypes.bool,
    portableUrl: PropTypes.string,
    thumbnail: PropTypes.string,
    url: PropTypes.string,
  }).isRequired,
  showId: PropTypes.bool,
};

export default GalleryCard;
