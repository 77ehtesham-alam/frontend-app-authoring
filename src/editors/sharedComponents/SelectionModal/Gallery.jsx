import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import {
  Scrollable, SelectableBox, Spinner,
} from '@edx/paragon';

import {
  FormattedMessage,
  injectIntl,
  intlShape,
  MessageDescriptor,
} from '@edx/frontend-platform/i18n';

import { selectors } from '../../data/redux';
import { RequestKeys } from '../../data/constants/requests';

import messages from './messages';
import GalleryCard from './GalleryCard';

export const Gallery = ({
  galleryIsEmpty,
  searchIsEmpty,
  displayList,
  highlighted,
  onHighlightChange,
  emptyGalleryLabel,
  showIdsOnCards,
  height,
  // injected
  intl,
  // redux
  isLoaded,
}) => {
  if (!isLoaded) {
    return (
      <Spinner
        animation="border"
        className="mie-3"
        screenReaderText={intl.formatMessage(messages.loading)}
      />
    );
  }
  if (galleryIsEmpty) {
    return (
      <div className="gallery p-4 bg-gray-100" style={{ height }}>
        <FormattedMessage {...emptyGalleryLabel} />
      </div>
    );
  }
  if (searchIsEmpty) {
    return (
      <div className="gallery p-4 bg-gray-100" style={{ height }}>
        <FormattedMessage {...messages.emptySearchLabel} />
      </div>
    );
  }
  return (
    <Scrollable className="gallery bg-gray-100" style={{ height }}>
      <div className="p-4">
        <SelectableBox.Set
          columns={1}
          name="images"
          onChange={onHighlightChange}
          type="radio"
          value={highlighted}
        >
          { displayList.map(asset => <GalleryCard key={asset.id} asset={asset} showId={showIdsOnCards} />) }
        </SelectableBox.Set>
      </div>
    </Scrollable>
  );
};

Gallery.defaultProps = {
  highlighted: '',
  showIdsOnCards: false,
  height: '375px',
  emptyGalleryLabel: null,
};
Gallery.propTypes = {
  galleryIsEmpty: PropTypes.bool.isRequired,
  searchIsEmpty: PropTypes.bool.isRequired,
  displayList: PropTypes.arrayOf(PropTypes.object).isRequired,
  highlighted: PropTypes.string,
  onHighlightChange: PropTypes.func.isRequired,
  emptyGalleryLabel: MessageDescriptor,
  showIdsOnCards: PropTypes.bool,
  height: PropTypes.string,
  // injected
  intl: intlShape.isRequired,
  // redux
  isLoaded: PropTypes.bool.isRequired,
};

const requestKey = RequestKeys.fetchAssets;
export const mapStateToProps = (state) => ({
  isLoaded: selectors.requests.isFinished(state, { requestKey }),
});

export const mapDispatchToProps = {};

export default injectIntl(connect(mapStateToProps, mapDispatchToProps)(Gallery));
