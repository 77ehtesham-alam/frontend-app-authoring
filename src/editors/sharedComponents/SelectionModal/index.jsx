import React from 'react';
import PropTypes from 'prop-types';

import { Button, Stack } from '@edx/paragon';
import { Add } from '@edx/paragon/icons';
import {
  FormattedMessage,
  injectIntl,
  MessageDescriptor,
  intlShape,
} from '@edx/frontend-platform/i18n';

import BaseModal from '../BaseModal';
import SearchSort from './SearchSort';
import Gallery from './Gallery';
import FileInput from '../FileInput';
import ErrorAlert from '../ErrorAlerts/ErrorAlert';
import FetchErrorAlert from '../ErrorAlerts/FetchErrorAlert';
import UploadErrorAlert from '../ErrorAlerts/UploadErrorAlert';

export const SelectionModal = ({
  isOpen,
  close,
  size,
  isFullscreenScroll,
  galleryError,
  inputError,
  fileInput,
  galleryProps,
  searchSortProps,
  selectBtnProps,
  acceptedFiles,
  modalMessages,
  // injected
  intl,
}) => {
  const {
    confirmMsg,
    uploadButtonMsg,
    titleMsg,
    fetchError,
    uploadError,
  } = modalMessages;
  return (
    <BaseModal
      close={close}
      confirmAction={(
        <Button {...selectBtnProps} variant="primary">
          <FormattedMessage {...confirmMsg} />
        </Button>
      )}
      isOpen={isOpen}
      size={size}
      isFullscreenScroll={isFullscreenScroll}
      footerAction={(
        <Button iconBefore={Add} onClick={fileInput.click} variant="link">
          <FormattedMessage {...uploadButtonMsg} />
        </Button>
      )}
      title={intl.formatMessage(titleMsg)}
    >
      {/* Error Alerts */}
      <FetchErrorAlert message={fetchError} />
      <UploadErrorAlert message={uploadError} />
      <ErrorAlert
        dismissError={inputError.dismiss}
        hideHeading
        isError={inputError.show}
      >
        <FormattedMessage {...inputError.message} />
      </ErrorAlert>

      {/* User Feedback Alerts */}
      <ErrorAlert
        dismissError={galleryError.dismiss}
        hideHeading
        isError={galleryError.show}
      >
        <FormattedMessage {...galleryError.message} />
      </ErrorAlert>
      <Stack gap={3}>
        <SearchSort {...searchSortProps} />
        <Gallery {...galleryProps} />
        <FileInput fileInput={fileInput} acceptedFiles={Object.values(acceptedFiles).join()} />
      </Stack>
    </BaseModal>
  );
};

SelectionModal.defaultProps = {
  size: 'lg',
  isFullscreenScroll: true,
};

SelectionModal.propTypes = {
  isOpen: PropTypes.bool.isRequired,
  close: PropTypes.func.isRequired,
  size: PropTypes.string,
  isFullscreenScroll: PropTypes.bool,
  galleryError: PropTypes.shape({
    dismiss: PropTypes.func.isRequired,
    show: PropTypes.bool.isRequired,
    set: PropTypes.func.isRequired,
    message: MessageDescriptor,
  }).isRequired,
  inputError: PropTypes.shape({
    dismiss: PropTypes.func.isRequired,
    show: PropTypes.bool.isRequired,
    set: PropTypes.func.isRequired,
    message: MessageDescriptor,
  }).isRequired,
  fileInput: PropTypes.shape({
    click: PropTypes.func.isRequired,
    addFile: PropTypes.func.isRequired,
  }).isRequired,
  galleryProps: PropTypes.shape({}).isRequired,
  searchSortProps: PropTypes.shape({}).isRequired,
  selectBtnProps: PropTypes.shape({}).isRequired,
  acceptedFiles: PropTypes.shape({}).isRequired,
  modalMessages: PropTypes.shape({
    confirmMsg: MessageDescriptor,
    uploadButtonMsg: MessageDescriptor,
    titleMsg: MessageDescriptor,
    fetchError: MessageDescriptor,
    uploadError: MessageDescriptor,
  }).isRequired,
  // injected
  intl: intlShape.isRequired,
};

export default injectIntl(SelectionModal);
