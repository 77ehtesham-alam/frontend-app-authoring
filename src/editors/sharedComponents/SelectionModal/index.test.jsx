import React from 'react';
import { shallow } from 'enzyme';
import { formatMessage } from '../../../testUtils';
import SelectionModal from '.';

const props = {
  isOpen: jest.fn(),
  isClose: jest.fn(),
  size: 'fullscreen',
  isFullscreenScroll: false,
  galleryError: {
    show: 'ShoWERror gAlLery',
    set: jest.fn(),
    dismiss: jest.fn(),
    message: {
      id: 'Gallery error id',
      defaultMessage: 'Gallery error',
      description: 'Gallery error',
    },
  },
  inputError: {
    show: 'ShoWERror inPUT',
    set: jest.fn(),
    dismiss: jest.fn(),
    message: {
      id: 'Input error id',
      defaultMessage: 'Input error',
      description: 'Input error',
    },
  },
  fileInput: {
    addFile: 'imgHooks.fileInput.addFile',
    click: 'imgHooks.fileInput.click',
    ref: 'imgHooks.fileInput.ref',
  },
  galleryProps: { gallery: 'props' },
  searchSortProps: { search: 'sortProps' },
  selectBtnProps: { select: 'btnProps' },
  acceptedFiles: { png: '.png' },
  modalMessages: {
    confirmMsg: {
      id: 'confirmMsg',
      defaultMessage: 'confirmMsg',
      description: 'confirmMsg',
    },
    uploadButtonMsg: {
      id: 'uploadButtonMsg',
      defaultMessage: 'uploadButtonMsg',
      description: 'uploadButtonMsg',
    },
    titleMsg: {
      id: 'titleMsg',
      defaultMessage: 'titleMsg',
      description: 'titleMsg',
    },
    fetchError: {
      id: 'fetchError',
      defaultMessage: 'fetchError',
      description: 'fetchError',
    },
    uploadError: {
      id: 'uploadError',
      defaultMessage: 'uploadError',
      description: 'uploadError',
    },
  },
  intl: { formatMessage },
};

jest.mock('../BaseModal', () => 'BaseModal');
jest.mock('./SearchSort', () => 'SearchSort');
jest.mock('./Gallery', () => 'Gallery');
jest.mock('../FileInput', () => 'FileInput');
jest.mock('../ErrorAlerts/ErrorAlert', () => 'ErrorAlert');
jest.mock('../ErrorAlerts/FetchErrorAlert', () => 'FetchErrorAlert');
jest.mock('../ErrorAlerts/UploadErrorAlert', () => 'UploadErrorAlert');

describe('Selection Modal', () => {
  describe('snapshots', () => {
    test('rendering correctly with expected Input', () => {
      expect(shallow(<SelectionModal {...props} />)).toMatchSnapshot();
    });
    test('rendering with props to null', () => {
      expect(shallow(<SelectionModal />)).toMatchSnapshot();
    });
  });
});
