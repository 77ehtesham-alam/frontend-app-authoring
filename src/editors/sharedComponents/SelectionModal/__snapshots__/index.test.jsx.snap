// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Selection Modal snapshots rendering correctly with expected Input 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;

exports[`Selection Modal snapshots rendering with props to null 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
